### ISSUE TITLE
Add an NLog logging integration

### ISSUE THREAD

**Issue by RudeySH** [getsentry/sentry-dotnet#167]
It would be a good addition next to the existing Log4Net and Serilog integrations. Currently, there are at least 7 NLog+Sentry NuGet packages made by third parties, and frankly I'm not sure how to decide which one to use and it's making it a bit harder for me to start using Sentry.

**Comment by bruno-garcia** [comment]
@RudeySH thanks for bringing this to us. 
The current NLog integrations, to my knowledge, are based on our legacy [Raven SDK](https://github.com/getsentry/raven-csharp). We plan to add an NLog integration but it's still a matter of prioritization with other projects within Sentry.

We'd be happy to merge a PR though! It'd be pretty similar to the Serilog integration so basically copy pasting the Serilog project (and tests/sample) and adapting to work with NLog instead.

**Comment by UgurAldanmaz** [comment]
Anyone working on this?

**Comment by bruno-garcia** [comment]
@UgurAldanmaz no one ATM. This is still up for grabs.

**Comment by josh-degraw** [comment]
I'd like to take this on if it's still up for grabs.

**Comment by bruno-garcia** [comment]
@josh-degraw It'd be great to get some Help. Thanks!

**Comment by josh-degraw** [comment]
Awesome. I'm working on it now, I've got most of the plumbing set up, are there any non-obvious design conventions that I should adhere to? I couldn't find and contributing guidelines, so so far I've just been kind of duplicating the existing projects.

**Comment by bruno-garcia** [comment]
Great! Thanks again. 
No special conventions besides what's in the .editorconfig

I think this comment is still relevant:

> We'd be happy to merge a PR though! It'd be pretty similar to the Serilog integration so basically copy pasting the Serilog project (and tests/sample) and adapting to work with NLog instead.

The log4net integration wouldn't be a good candidate since it isn't structured logging and doesn't add breadcrumbs as the Serilog does.

**Comment by josh-degraw** [comment]
Yeah that's basically what I've done so far. I've got it working, I'm just running into an issue (that appears to be on NLog's side) where it's not disposing of the target when it's supposed to, so the messages don't get sent unless the target is explicitly disposed.
Once I can get that figured out I'll send a PR.

**Comment by snakefoot** [comment]
NLog has these target-state methods:

- InitializeTarget -> Should create sentry-sdk-client
- FlushAsync
- CloseTarget -> Should dispose sentry-sdk-client

NLog performs parallel flush of all targets, and performs sequential dispose of all targets. This means that if a NLog-target is very slow to dispose/close (network traffic/timeouts etc.) then it will seriously affect the shutdown of the remaining NLog-targets.

For optimal behavior then sdk-client should have a Flush-method that can be called by FlushAsync (log4net-appenders also likes to Flush).

**Comment by snakefoot** [comment]
@josh-degraw I can recommend that you base your SentryNLogTarget on `NLog.Targets.TargetWithContext` (Introduced with NLog 4.5).

Then just call the method:

`Dictionary<string,object> allProperties = this.GetAllProperties();`

**Comment by josh-degraw** [comment]
@snakefoot thanks that was it! I was doing the cleanup in `Dispose()` where I thought it needed to be. Put it in `CloseTarget()` and it looks like it worked. Just need to clean up a bit and I'll push the PR. Also, yeah I was looking for flush-like behavior in Sentry but couldn't find where to do that if there is a way already.

**Comment by bruno-garcia** [comment]
We're pending a Flush method on the sdk. We'll get one in soon.

**Comment by bruno-garcia** [comment]
We've released the first beta (Thanks a lot @josh-degraw !)

Please try it out and let us know how it works :)
https://www.nuget.org/packages/Sentry.NLog/1.1.3-beta3

**Comment by josh-degraw** [comment]
Thanks @bruno-garcia! Glad to be able to contribute!

**Comment by RudeySH** [comment]
Big thanks to everyone involved! This integration makes Sentry a viable option for me.

Is it possible to prevent the integration from initializing the SDK when I'm using XML to configure NLog? The docs specify that not providing a DSN prevents the integration from initializing the SDK, but this doesn't seem to be true when configuring NLog using XML.

    Required parameter 'Dsn' on 'Sentry Target[Sentry]' was not specified.

I also tried using `initializeSdk="false"`, but that doesn't seem to work either.

    Error when setting 'false' on attibute 'initializeSdk' Exception: System.NotSupportedException: Parameter initializeSdk not supported on SentryTarget

I want to continue using XML to configure NLog, but I also want to use `SentrySdk.Init` to initialize Sentry, because I'm also using `options.AddEntityFramework()` there.

**Comment by josh-degraw** [comment]
@RudeySH Yes, this was something I realized should be changed after this pull request was merged, but I have already submitted a PR (#205) to fix this, so once that's merged it should be fixed in the next release.

Also, as for specifying `initializeSdk="false"`, for now that should be set like this:

```xml
<target type="Sentry">
    <options initializeSdk="false" />
</target
```

Although I think I'm going to submit a PR that brings those up to the `target` level rather than directly setting it on the property, because the documentation on that is kind of confusing and it's really less "standard" as far as NLog configs go anyways.

**Comment by josh-degraw** [comment]
Okay, since it hadn't been merged yet and it was a pretty quick fix, I just went ahead and closed the other PR and submitted #207 which does both of these (unmarking dsn as required in the xml and pulling up the properties from `Options` onto the `SentryTarget`. 

More advanced options only declared on the base class `SentryOptions` can still be configured in the above manner, however, but those will only take effect when this integration is the one initializing the sdk.

**Comment by bruno-garcia** [comment]
@josh-degraw thanks for the PR. It's merged and publish on version `1.2.1-beta`.
I'll close this issue. Please raise a new one to track new workloads related to NLog.
